In [19]:
import numpy as np
np.median(np.array([[0.00299,2.06,125.0,0.0177,10.4,122.0,0.0093,10.9,84.1],[0.00631,2.53,125.0,0.00844,4.67,126.0,0.0103,6.3,80.9],[0.00465,3.92,109.0,0.0136,12.0,124.0,0.00498,2.13,79.6]]), axis=0)

array([4.65e-03, 2.53e+00, 1.25e+02, 1.36e-02, 1.04e+01, 1.24e+02,
       9.30e-03, 6.30e+00, 8.09e+01])

### Testing data cleaning codes on one json item

In [1]:
# import json
# import numpy as np
# import pandas as pd
# with open('data_test.json') as train_file:
#     data = json.load(train_file)

In [2]:
# transcript_name = []
# position_num = []
# bases_corresponding_to_position = []
# df = pd.DataFrame(columns=['Transcript Name', 'Position', 'Bases', 'SD', 'Mean', 'Dwelling Time'])
# for k, v in data.items(): # obtain transcript name and rest of the items
#     print(k)
#     print(v)
#     transcript_name.append(k)
#     transcript_name = transcript_name * 3
#     transcript_name = pd.DataFrame(transcript_name, columns=['Transcript Name'])
#     print('')
#     for i,j in v.items(): # obtain position and rest of items
#         print(i)
#         print(j)
#         position_num.append(int(i) - 1)
#         position_num.append(int(i))
#         position_num.append(int(i) + 1)
#         position_num = pd.DataFrame(position_num, columns = ['Position'])
#         print('')
#         for a,b in j.items(): # obtain bases and measurements - dwell time, etc.
#             print("a", a)
#             print("b", b)
#             bases_corresponding_to_position.append(a[:5])
#             bases_corresponding_to_position.append(a[1:6])
#             bases_corresponding_to_position.append(a[2:])
#             bases_corresponding_to_position = pd.DataFrame(bases_corresponding_to_position, columns = ['Bases'])
            
#             arr = np.array(b)
#             mean_measurement_res = np.mean(arr, axis=0)
#             print(mean_measurement_res)
#             measurement_first_pos = mean_measurement_res[0:3]
#             measurement_second_pos = mean_measurement_res[3:6]
#             measurement_third_pos = mean_measurement_res[6:]
            
#             measurement_first_pos = measurement_first_pos.reshape(1,3)
#             measurement_second_pos = measurement_second_pos.reshape(1,3)
#             measurement_third_pos = measurement_third_pos.reshape(1,3)
#             measurement_first_pos = pd.DataFrame.from_records(measurement_first_pos, columns = ['Dwelling Time', 'SD', 'Mean'])
#             measurement_second_pos = pd.DataFrame.from_records(measurement_second_pos, columns = ['Dwelling Time', 'SD', 'Mean'])
#             measurement_third_pos = pd.DataFrame.from_records(measurement_third_pos, columns = ['Dwelling Time', 'SD', 'Mean'])
            
#             all_measurement = pd.concat([measurement_first_pos, measurement_second_pos, measurement_third_pos], axis=0).reset_index(drop=True)
#             full_df = pd.concat([transcript_name, position_num, bases_corresponding_to_position, all_measurement], axis=1)
            
#             print(full_df)

ENST00000000233
{'244': {'AAGACCA': [[0.00299, 2.06, 125.0, 0.0177, 10.4, 122.0, 0.0093, 10.9, 84.1], [0.00631, 2.53, 125.0, 0.00844, 4.67, 126.0, 0.0103, 6.3, 80.9], [0.00465, 3.92, 109.0, 0.0136, 12.0, 124.0, 0.00498, 2.13, 79.6], [0.00398, 2.06, 125.0, 0.0083, 5.01, 130.0, 0.00498, 3.78, 80.4], [0.00664, 2.92, 120.0, 0.00266, 3.94, 129.0, 0.013, 7.15, 82.2], [0.0103, 3.83, 123.0, 0.00598, 6.45, 126.0, 0.0153, 1.09, 74.8], [0.00398, 3.75, 126.0, 0.00332, 4.3, 129.0, 0.00299, 1.93, 81.9], [0.00498, 3.93, 127.0, 0.00398, 2.51, 131.0, 0.0111, 3.47, 79.4], [0.0139, 4.69, 106.0, 0.0136, 6.21, 124.0, 0.00531, 10.6, 85.5], [0.00631, 3.5, 126.0, 0.0222, 5.38, 128.0, 0.00332, 1.72, 79.3], [0.0061, 3.99, 121.0, 0.0121, 7.27, 122.0, 0.00232, 1.27, 78.9], [0.00299, 1.99, 128.0, 0.00427, 4.85, 124.0, 0.00332, 3.18, 80.5], [0.0186, 3.62, 124.0, 0.00428, 2.25, 129.0, 0.00554, 2.78, 80.1], [0.0093, 3.12, 125.0, 0.00398, 8.84, 129.0, 0.00361, 1.86, 82.0], [0.00365, 2.92, 126.0, 0.00698, 3.7, 126.0, 0

### Data cleaning codes for all the json items

Read measurements are given in the following form: `[0.00299,2.06,125.0,0.0177,10.4,122.0,0.0093,10.9,84.1]`. We can interpret the numbers as such:
- 1st, 4th and 7th item refers to the dwelling time of the 5-mer nucleotides
- 2nd, 5th and 8th item refers to the standard deviation of the direct RNA current 
- 3rd, 6th and 9th item refers to the mean of the direct RNA current

For each read we have extracted features from segments corresponding to 1-neighboring position and this results in a total of 9 features for each read (3 features for position 243, followed by 3 features for positions 244 and 245 respectively).

Since there are variable number of reads for each position, will be trying to obtain the **mean/ median** values across all the reads. E.g., if the measurements are `[0.00299,2.06,125.0,0.0177,10.4,122.0,0.0093,10.9,84.1],[0.00631,2.53,125.0,0.00844,4.67,126.0,0.0103,6.3,80.9],[0.00465,3.92,109.0,0.0136,12.0,124.0,0.00498,2.13,79.6]`, the result I will be using is `[4.65e-03, 2.53e+00, 1.25e+02, 1.36e-02, 1.04e+01, 1.24e+02, 9.30e-03, 6.30e+00, 8.09e+01]`

In [1]:
import json
import numpy as np
import pandas as pd
data_list = [json.loads(line)
        for line in open('./Data/data.json', 'r', encoding='utf-8')]

In [2]:
df = pd.DataFrame(columns=['Transcript Name', 'Position', 'Bases', 'Mean SD', 'Mean_Mean', 'Mean Dwelling Time', 'Median SD', 'Median_Mean', 'Median Dwelling Time'])
for data in data_list:
    transcript_name = []
    position_num = []
    bases_corresponding_to_position = []
    for k, v in data.items(): # obtain transcript name and rest of the items
        transcript_name.append(k)
        transcript_name = transcript_name * 3
        transcript_name = pd.DataFrame(transcript_name, columns=['Transcript Name'])
        for i,j in v.items(): # obtain position and rest of items
            position_num.append(int(i) - 1)
            position_num.append(int(i))
            position_num.append(int(i) + 1)
            position_num = pd.DataFrame(position_num, columns = ['Position'])
            for a,b in j.items(): # obtain bases and measurements - dwell time, etc.
                bases_corresponding_to_position.append(a[:5])
                bases_corresponding_to_position.append(a[1:6])
                bases_corresponding_to_position.append(a[2:])
                bases_corresponding_to_position = pd.DataFrame(bases_corresponding_to_position, columns = ['Bases'])

                arr = np.array(b)
                
                # For median readings
                median_measurement_res = np.median(arr, axis=0)
                median_measurement_first_pos = median_measurement_res[0:3]
                median_measurement_second_pos = median_measurement_res[3:6]
                median_measurement_third_pos = median_measurement_res[6:]

                median_measurement_first_pos = median_measurement_first_pos.reshape(1,3)
                median_measurement_second_pos = median_measurement_second_pos.reshape(1,3)
                median_measurement_third_pos = median_measurement_third_pos.reshape(1,3)
                median_measurement_first_pos = pd.DataFrame.from_records(median_measurement_first_pos, columns = ['Median Dwelling Time', 'Median SD', 'Median_Mean'])
                median_measurement_second_pos = pd.DataFrame.from_records(median_measurement_second_pos, columns = ['Median Dwelling Time', 'Median SD', 'Median_Mean'])
                median_measurement_third_pos = pd.DataFrame.from_records(median_measurement_third_pos, columns = ['Median Dwelling Time', 'Median SD', 'Median_Mean'])

                all_median_measurement = pd.concat([median_measurement_first_pos, median_measurement_second_pos, median_measurement_third_pos], axis=0).reset_index(drop=True)
                
                # For mean readings
                mean_measurement_res = np.mean(arr, axis=0)
                measurement_first_pos = mean_measurement_res[0:3]
                measurement_second_pos = mean_measurement_res[3:6]
                measurement_third_pos = mean_measurement_res[6:]

                measurement_first_pos = measurement_first_pos.reshape(1,3)
                measurement_second_pos = measurement_second_pos.reshape(1,3)
                measurement_third_pos = measurement_third_pos.reshape(1,3)
                measurement_first_pos = pd.DataFrame.from_records(measurement_first_pos, columns = ['Mean Dwelling Time', 'Mean SD', 'Mean_Mean'])
                measurement_second_pos = pd.DataFrame.from_records(measurement_second_pos, columns = ['Mean Dwelling Time', 'Mean SD', 'Mean_Mean'])
                measurement_third_pos = pd.DataFrame.from_records(measurement_third_pos, columns = ['Mean Dwelling Time', 'Mean SD', 'Mean_Mean'])

                all_measurement = pd.concat([measurement_first_pos, measurement_second_pos, measurement_third_pos], axis=0).reset_index(drop=True)
                
                full_df = pd.concat([transcript_name, position_num, bases_corresponding_to_position, all_measurement, all_median_measurement], axis=1)

                df = df.append(full_df).reset_index(drop=True)

In [3]:
# df.to_csv('./Data/full_df_2.csv', index=False)

In [4]:
df

Transcript Name Position  Bases   Mean SD   Mean_Mean  \
0       ENST00000000233      243  AAGAC  4.223784  123.702703   
1       ENST00000000233      244  AGACC  7.382162  125.913514   
2       ENST00000000233      245  GACCA  4.386989   80.570270   
3       ENST00000000233      260  CAAAC  3.216424  109.681395   
4       ENST00000000233      261  AAACT  3.226535  107.889535   
...                 ...      ...    ...       ...         ...   
365509  ENST00000641834     1537  TGACC  6.552982  123.263158   
365510  ENST00000641834     1538  GACCA  2.540877   82.289474   
365511  ENST00000641834     1692  TTGAC  4.090577  105.807692   
365512  ENST00000641834     1693  TGACA  8.702885  113.134615   
365513  ENST00000641834     1694  GACAT  2.576731   78.536538   

        Mean Dwelling Time  Median SD  Median_Mean  Median Dwelling Time  
0                 0.008264      3.730       125.00              0.006970  
1                 0.009373      6.650       126.00              0.007970  
2                 0.007345      3.440        80.50              0.005980  
3                 0.006609      2.880       110.00              0.005640  
4                 0.006813      3.000       108.00              0.005885  
...                    ...        ...          ...                   ...  
365509            0.007419      5.790       124.00              0.006810  
365510            0.006472      2.330        82.00              0.006310  
365511            0.008788      3.160       107.00              0.007090  
365512            0.006907      8.675       113.00              0.006705  
365513            0.008337      2.295        77.45              0.006205  

[365514 rows x 9 columns]

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365514 entries, 0 to 365513
Data columns (total 9 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   Transcript Name       365514 non-null  object 
 1   Position              365514 non-null  object 
 2   Bases                 365514 non-null  object 
 3   Mean SD               365514 non-null  float64
 4   Mean_Mean             365514 non-null  float64
 5   Mean Dwelling Time    365514 non-null  float64
 6   Median SD             365514 non-null  float64
 7   Median_Mean           365514 non-null  float64
 8   Median Dwelling Time  365514 non-null  float64
dtypes: float64(6), object(3)
memory usage: 25.1+ MB


In [5]:
labels = pd.read_csv('./Data/data.info')
full_df = df.merge(labels, how='left', left_on = ['Transcript Name', 'Position'], right_on = ['transcript_id', 'transcript_position'])
full_df

Transcript Name Position  Bases   Mean SD   Mean_Mean  \
0       ENST00000000233      243  AAGAC  4.223784  123.702703   
1       ENST00000000233      244  AGACC  7.382162  125.913514   
2       ENST00000000233      245  GACCA  4.386989   80.570270   
3       ENST00000000233      260  CAAAC  3.216424  109.681395   
4       ENST00000000233      261  AAACT  3.226535  107.889535   
...                 ...      ...    ...       ...         ...   
365509  ENST00000641834     1537  TGACC  6.552982  123.263158   
365510  ENST00000641834     1538  GACCA  2.540877   82.289474   
365511  ENST00000641834     1692  TTGAC  4.090577  105.807692   
365512  ENST00000641834     1693  TGACA  8.702885  113.134615   
365513  ENST00000641834     1694  GACAT  2.576731   78.536538   

        Mean Dwelling Time  Median SD  Median_Mean  Median Dwelling Time  \
0                 0.008264      3.730       125.00              0.006970   
1                 0.009373      6.650       126.00              0.007970   
2                 0.007345      3.440        80.50              0.005980   
3                 0.006609      2.880       110.00              0.005640   
4                 0.006813      3.000       108.00              0.005885   
...                    ...        ...          ...                   ...   
365509            0.007419      5.790       124.00              0.006810   
365510            0.006472      2.330        82.00              0.006310   
365511            0.008788      3.160       107.00              0.007090   
365512            0.006907      8.675       113.00              0.006705   
365513            0.008337      2.295        77.45              0.006205   

                gene_id    transcript_id  transcript_position  label  
0                   NaN              NaN                  NaN    NaN  
1       ENSG00000004059  ENST00000000233                244.0    0.0  
2                   NaN              NaN                  NaN    NaN  
3                   NaN              NaN                  NaN    NaN  
4       ENSG00000004059  ENST00000000233                261.0    0.0  
...                 ...              ...                  ...    ...  
365509  ENSG00000167747  ENST00000641834               1537.0    0.0  
365510              NaN              NaN                  NaN    NaN  
365511              NaN              NaN                  NaN    NaN  
365512  ENSG00000167747  ENST00000641834               1693.0    0.0  
365513              NaN              NaN                  NaN    NaN  

[365514 rows x 13 columns]

In [7]:
full_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 365514 entries, 0 to 365513
Data columns (total 13 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   Transcript Name       365514 non-null  object 
 1   Position              365514 non-null  object 
 2   Bases                 365514 non-null  object 
 3   Mean SD               365514 non-null  float64
 4   Mean_Mean             365514 non-null  float64
 5   Mean Dwelling Time    365514 non-null  float64
 6   Median SD             365514 non-null  float64
 7   Median_Mean           365514 non-null  float64
 8   Median Dwelling Time  365514 non-null  float64
 9   gene_id               121838 non-null  object 
 10  transcript_id         121838 non-null  object 
 11  transcript_position   121838 non-null  float64
 12  label                 121838 non-null  float64
dtypes: float64(8), object(5)
memory usage: 39.0+ MB


In [8]:
full_df[full_df['Transcript Name'] == 'ENST00000000233']

Transcript Name Position  Bases   Mean SD   Mean_Mean  Mean Dwelling Time  \
0   ENST00000000233      243  AAGAC  4.223784  123.702703            0.008264   
1   ENST00000000233      244  AGACC  7.382162  125.913514            0.009373   
2   ENST00000000233      245  GACCA  4.386989   80.570270            0.007345   
3   ENST00000000233      260  CAAAC  3.216424  109.681395            0.006609   
4   ENST00000000233      261  AAACT  3.226535  107.889535            0.006813   
5   ENST00000000233      262  AACTG  3.016599   94.290698            0.007710   
6   ENST00000000233      315  GAAAC  2.940541  105.475676            0.007570   
7   ENST00000000233      316  AAACA  3.642703   98.947027            0.007416   
8   ENST00000000233      317  AACAG  2.087146   89.364324            0.007555   
9   ENST00000000233      331  AGAAC  6.476350  129.355000            0.010620   
10  ENST00000000233      332  GAACA  2.899200   97.836500            0.008632   
11  ENST00000000233      333  AACAT  2.236520   89.154000            0.006102   
12  ENST00000000233      367  AGGAC  6.415051  117.924242            0.010701   
13  ENST00000000233      368  GGACA  5.870303  121.954545            0.011479   
14  ENST00000000233      369  GACAA  4.260253   85.178788            0.010019   
15  ENST00000000233      403  AGAAC  6.385260  128.046875            0.009011   
16  ENST00000000233      404  GAACA  2.938281   96.858333            0.008555   
17  ENST00000000233      405  AACAC  2.056797   91.591146            0.005798   
18  ENST00000000233      430  TGGAC  3.348642  116.864198            0.008683   
19  ENST00000000233      431  GGACA  5.648333  121.345679            0.007249   
20  ENST00000000233      432  GACAG  2.731636   85.971605            0.007839   
21  ENST00000000233      439  ATGAC  4.521182   96.130542            0.006788   
22  ENST00000000233      440  TGACC  9.343350  119.669951            0.008527   
23  ENST00000000233      441  GACCG  3.505099   81.061084            0.007414   
24  ENST00000000233      470  TGAAC  8.128495  120.597087            0.006060   
25  ENST00000000233      471  GAACT  3.058350  101.607282            0.006997   
26  ENST00000000233      472  AACTC  1.938034   92.542233            0.006358   
27  ENST00000000233      538  AGGAC  5.953696  116.989130            0.011070   
28  ENST00000000233      539  GGACA  7.052391  119.538043            0.008996   
29  ENST00000000233      540  GACAT  3.482826   81.005978            0.006398   
30  ENST00000000233      570  CTGAC  3.965594  109.933663            0.006936   
31  ENST00000000233      571  TGACT  8.138020  124.054455            0.006799   
32  ENST00000000233      572  GACTG  3.067327   87.861881            0.010206   
33  ENST00000000233      574  CTGAC  3.922727  108.110227            0.006600   
34  ENST00000000233      575  TGACA  8.337102  115.340909            0.007055   
35  ENST00000000233      576  GACAA  3.102045   80.612500            0.006868   
36  ENST00000000233      664  TGGAC  2.750854  120.517073            0.007405   
37  ENST00000000233      665  GGACT  5.511122  124.917073            0.008447   
38  ENST00000000233      666  GACTG  3.430976   89.932195            0.007119   
39  ENST00000000233      695  CTAAC  2.248129   91.996535            0.006931   
40  ENST00000000233      696  TAACC  3.112129   94.114851            0.007476   
41  ENST00000000233      697  AACCA  3.785995   83.872772            0.005992   
42  ENST00000000233      751  ATGAC  4.992540   87.989418            0.007023   
43  ENST00000000233      752  TGACC  8.945344  116.455026            0.008075   
44  ENST00000000233      753  GACCA  3.549259   76.803175            0.005891   
45  ENST00000000233      756  CAGAC  4.842452  110.600000            0.007799   
46  ENST00000000233      757  AGACT  6.047837  128.836538            0.010213   
47  ENST00000000233      758  GACTC  4.393942   90.439904            0.013364   
48  ENST00000000233      764  CGGAC